DTLS 1.3: Add dtls13_prepare_record_header

fj-blanco opened the pull request:
Implemented dtls13_prepare_record_header() function for writing the new DTLS 1.3 unified header to DTLS 1.3 DTLSCiphertext records.

Sequence number and length fields are currently hardcoded to use 16-bit representations. Connection ID is not yet supported.

The unified header writing is conditionally called when record type is SSL3_RT_APPLICATION_DATA.
Future work should verify the conditional logic for invoking dtls13_prepare_unified_header to ensure just DTLS 1.3 DTLSCipertext records receive the unified header.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

fj-blanco committed:
DTLS 1.3: Add dtls13_prepare_record_header
Implemented dtls13_prepare_record_header() function for writing the
new DTLS 1.3 unified header to DTLS 1.3 DTLSCiphertext records.

Sequence number and length fields are currently hardcoded to use
16-bit representations. Connection ID is not yet supported.

The unified header writing is conditionally called when record type
is SSL3_RT_APPLICATION_DATA.
Future work should verify the conditional logic for invoking
dtls13_prepare_unified_header to ensure just DTLS 1.3 DTLSCipertext
records receive the unified header.

fwh-dc commented on ssl/record/methods/dtls13_meth.c:
```diff
@@ -0,0 +1,83 @@
+
+    if (rectype == SSL3_RT_APPLICATION_DATA) {
+    /* DTLSCiphertext*/
+    /* TODO (DTLSv1.3): Ensure that we always use SSL3_RT_APPLICATION_DATA when encrypting in DTLS 1.3
```
Fix todo and indentation of this block :)

fwh-dc commented on ssl/record/methods/dtls13_meth.c:
```diff
@@ -0,0 +1,83 @@
+                               WPACKET *thispkt,
+                               OSSL_RECORD_TEMPLATE *templ,
+                               uint8_t rectype,
+                               unsigned char **recdata)
```
Fix indentation of args

fwh-dc commented on ssl/record/methods/dtls13_meth.c:
```diff
@@ -0,0 +1,83 @@
+        || (maxcomplen > 0
+            && !WPACKET_reserve_bytes(thispkt, maxcomplen,
+                                    recdata))) {
+        goto err;
```
Drop goto and simply:
```
    RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
    return 0;
```

fwh-dc commented on ssl/record/methods/dtls13_meth.c:
```diff
@@ -0,0 +1,83 @@
+
+err:
+    RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+    return 0;
```
remove

fwh-dc reviewed (commented):
Nice work!!

Now that you're familiar with the codebase, do you dare to also add the decoding of the headers?

fj-blanco commented:
> Nice work!!
> 
> Now that you're familiar with the codebase, do you dare to also add the decoding of the headers?

Great, thanks. I'll try!

fj-blanco commented:
> Nice work!!
> 
> Now that you're familiar with the codebase, do you dare to also add the decoding of the headers?

To confirm this is your idea. If we want to set the AAD header inside `tls13_cipher()` without introducing extra arguments, I think the only way is to read the write buffer of the `OSSL_RECORD_LAYER` object. Given that there's only one record present, we can obtain the first byte in `rl->wbuf[0].buf[offset]`, and we could decode the unified header with something like:

```c
/* Set up the AAD */
if (rl->isdtls) {
	//DTLS1.3 DTLSCiphertext record
	size_t offset = rl->wbuf[0].offset;
	unsigned char first_byte = rl->wbuf[0].buf[offset];

	if (!WPACKET_init_static_len(&wpkt, recheader, sizeof(recheader), 0)
	    || !WPACKET_put_bytes_u8(&wpkt, first_byte)
	    || !WPACKET_put_bytes_u8(&wpkt, rl->sequence[SEQ_NUM_SIZE - 2])
	    || !((first_byte & DTLS13_SBIT) ? WPACKET_put_bytes_u8(&wpkt, rl->sequence[SEQ_NUM_SIZE - 1]) : 1)
	    || !((first_byte & DTLS13_LBIT) ? WPACKET_put_bytes_u16(&wpkt, rec->length) : 1)
	    || !WPACKET_get_total_written(&wpkt, &hdrlen)
	    || hdrlen > DTLS13_UNIFIED_HEADER_MAX_LENGTH
	    || hdrlen < 2
	    || !WPACKET_finish(&wpkt)) {
	    RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
	    WPACKET_cleanup(&wpkt);
	    return 0;
	}

} else {
	//TLSCiphertext record
	[...]
}
```

Do you think this is ok?

fwh-dc commented:
I think it is better to hardcode it for now. We know that we will always be using the same first byte of the header when writing.

fj-blanco committed:
Fix DTLS 1.3 header handling, centralize defines

Resolved feedback on dtls13_prepare_record_header(). Moved DTLS 1.3
header defines to include/openssl/dtls1.h for shared use in
tls13_cipher() and future header parsing. Defines include:

- DTLS13_CBIT
- DTLS13_SBIT
- DTLS13_LBIT
- DTLS13_FIXED_BITS
- DTLS13_EPOCH_MASK

fj-blanco committed:
Adjust write buffer setup for DTLS 1.3 header length

Incorporate DTLS 1.3 specific header length adjustments across buffer
setup and initialization functions.

fj-blanco committed:
Implement AAD setup for DTLS 1.3 in tls13_cipher()
